src/org/rust/lang/core/psi/util/RustModExtensions.kt

```diff
@@ -0,0 +1,56 @@
+package org.rust.lang.core.psi.util
+
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RustModItem
+
+
+private val RustModItem.modDir: PsiDirectory?
+    get() {
+        val parent = containingMod ?: return containingFile.parent
+
+        val name = this.name ?: return null
+
+        return parent.modDir?.findSubdirectory(name)
+    }
+
+
+private val RustModItem.isCrateRoot: Boolean
+    get() = containingMod == null &&
+        (containingFile.name == "main.rs" || containingFile.name == "lib.rs")
+
+private val RustModItem.ownsDirectory: Boolean
+    get() = containingMod != null || // any inline nested module owns a directory
+        containingFile.name == "mod.rs" ||
+        isCrateRoot
+
+
+sealed class ChildModFile {
+    val mod: RustModItem?
+        get() = (this as? Found)?.file?.firstChild as? RustModItem
+
+    class NotFound : ChildModFile()
+    class Found(val file: PsiFile) : ChildModFile()
+    class Ambigious(val paths: Collection<PsiFile>) : ChildModFile()
+}
+
+fun RustModItem.submoduleFile(modName: String): ChildModFile {
+    if (!ownsDirectory) {
+        return ChildModFile.NotFound()
+    }
+
+    val dir = modDir
+    val dirMod = dir?.findSubdirectory(modName)?.findFile("mod.rs")
```

Comment:
Although I'm not happy with `RUST_DIRECTORY_MOD_NAME` constant name, but I failed to invent a better one (MOD_RS seems stupid, or is it Ok?)